### ISSUE TITLE
The test suite fails on Python 3.10: issue with factorial() on NaN

### ISSUE THREAD

**Issue by vstinner** [scipy/scipy#13122]
On Python 3.10, test_mixed_nan_inputs() of scipy.special.tests.test_basic.TestFactorialFunctions fails with:

```
                if un.size:
>                   val = math.factorial(un[0])
E                   TypeError: 'float' object cannot be interpreted as an integer
```

See the downstream Fedora issue: https://bugzilla.redhat.com/show_bug.cgi?id=1898157

I wrote PR #13121 to fix this issue.

The issue was worked around by the commit f7b6202ae69282b0942a2f5671737a85a558e1b9 for Python 3.9:

```
commit f7b6202ae69282b0942a2f5671737a85a558e1b9
Author: Ralf Gommers <ralf.gommers@gmail.com>
Date:   Sat May 9 15:31:07 2020 +0200

    TST: filter out deprecation warning for using factorial() with floats
    
    This was failing in py3.9-dev CI
```

cc @rgommers

#### Scipy/Numpy/Python version information:

* scipy  1.5.4-1.fc33 
* Python 3.10.0a2
* python3-numpy 1:1.19.4-2.fc34

**Comment by rgommers** [comment]
For 1.4.0, gh-11022 discusses a change in behaviour. Comments there indicate propagating `nan`'s, but no one pointed out that the function is not documented to accept floats to begin with.

The PR that closed that issue (PR gh-11254) does discuss it a bit, quoting: _I'm not opposed to forcing input `n` to be `int` in `_factorial`, but if we do that I think it should be consistent behavior within the function. For example, currently, `math.factorial(5.1)` throws a `ValueError` but `math.factorial(5.0)` is the correct `int` returned._

gh-11579 and gh-11984 both have discussion on deprecating `nan` input to `factorial`, that seems to be preferred. Unclear why gh-11984 was closed, there's no PR referring to it. Probably due to my commit f7b6202ae6928, but that was just a quick fix to stop CI from failing.

Cc @rlucas7 @person142 @andyfaff who worked on this before.

Random side note, searching history for `factorial` we also find this little gem in `scipy._lib._util`:
```
def float_factorial(n: int) -> float:
    """Compute the factorial and return as a float

    Returns infinity when result is too large for a double
    """
    return float(math.factorial(n)) if n < 171 else np.inf
```

Also, there's `factorial2` and `factorialk` functions which do not special-case `nan`.

Proposed solution:

0. Desired behaviour is to raise an exception on float input (_all_ float input, no special-casing of `[1.0, 2.0, 3.0]`).
1. In 1.6.0, (a) give a `DeprecationWarning` for float input, and (b) simply skip the problematic test for Python >=3.10
2. For 1.8.0, change to the desired behaviour.

**Comment by peterbell10** [comment]
>For 1.4.0, gh-11022 discusses a change in behaviour. Comments there indicate propagating nan's, but no one pointed out that the function is not documented to accept floats to begin with.

`factorial` is also documented as returning `int` when `exact=True` (or `object` to avoid overflow) but in the presence of NaNs it returns a `float` array instead. The result is no longer exact and the `float` type will overflow for any `n >= 171`.

**Comment by vstinner** [comment]
> factorial is also documented as returning int when exact=True (or object to avoid overflow) but in the presence of NaNs it returns a float array instead.

Feedback from someone completly new to numpy and scipy: I was surprised that putting a single NaN in an array of integers convert the whole array to floats. And yeah, I was confused by exact=True meaning :-) I would be less surprising if the function reject an array (raise an error) if a single value is a float if exact=True.

> In 1.6.0, (a) give a DeprecationWarning for float input, and (b) simply skip the problematic test for Python >=3.10

math.factorial() emits a DeprecationWarning on Python 3.9 if the argument is a float. See the scipy unit test which ignores the DeprecationWarning on purpose.

Yeah, skipping the whole test on Python 3.10 was my first attempt :-)

**Comment by andyfaff** [comment]
> Feedback from someone completly new to numpy and scipy: I was surprised that putting a single NaN in an array of integers convert the whole array to floats

```
>>> isinstance(np.nan, float)
True
```

numpy arrays are all of a single type (unless they're object arrays), so you can't mix int and float. The presence of a float in what is otherwise a list of ints coerces everything to a float.

**Comment by vstinner** [comment]
I wrote PR #13130 trivial change to only skip the discussed test on Python 3.10. It would unblock me for now, and then you can decide what to do in the long term ;-)

**Comment by vstinner** [comment]
> numpy arrays are all of a single type (unless they're object arrays), so you can't mix int and float. The presence of a float in what is otherwise a list of ints coerces everything to a float.

Sure. I understand the rationale. It's just to say that surprising that special.factorial() accepts NaN for exact=True. It looks like a bug in the API, unexpected special case.

It looks like it was a good idea that math.factorial() stops accepting float, so it forces thinking about the API that you want to provide ;-)

**Comment by rgommers** [comment]
Skipping the test on Python >= 3.10 was done in gh-13130. Reopening this issue because we still have to deprecate the current behaviour.

**Comment by h-vetinari** [comment]
Needs to be re-milestoned

**Comment by NeilGirdhar** [comment]
Is this issue the reason that SciPy's upper bound version is [set](https://github.com/scipy/scipy/blob/master/setup.py#L555) to be "<3.10"?  Because of that limit, poetry refuses to upgrade SciPy on any project that leaves the Python version unbounded.  Poetry says:
```
   1: fact: scipy (1.7.0) requires Python >=3.7,<3.10
   1: derived: not scipy (==1.7.0)
   1: fact: scipy (1.6.3) requires Python >=3.7,<3.10
   1: derived: not scipy (==1.6.3)
   1: fact: scipy (1.6.2) requires Python >=3.7,<3.10
   1: derived: not scipy (==1.6.2)
   1: fact: scipy (1.6.0) depends on numpy (>=1.16.5)
   1: selecting scipy (1.6.0)
```

**Comment by rgommers** [comment]
That's unrelated to this issue. Python 3.10 is simply not yet released and 1.7.0 therefore does not support it. Why is Poetry trying to use Python 3.10 already?

> Because of that limit, poetry refuses to upgrade SciPy on any project that leaves the Python version unbounded.

This seems like bad design anyway on Poetry's side. After Python 3.10.0 is released, it takes a while for projects to put up wheels that support that new Python version. So at that point Poetry will refuse to work for pretty much anyone that does not have an upper bound on Python version.

> I noticed that NumPy doesn't seem to do have this behaviour.

It will after I fix it. Currently `setup.py` says `python_requires='>=3.7',`, which is wrong on the 1.21.x maintenance branch.

**Comment by NeilGirdhar** [comment]
@rgommers Okay, I trust you :smi.   I hope you don't mind that I'll link your comment to the issue that I'll file with Poetry.
